contracts/dtoken/src/rewards.rs

```diff
@@ -0,0 +1,841 @@
+        self.reward_campaigns.get(&campaign_id)
+    }
+
+    pub fn get_reward_campaigns_extended(&self) -> Vec<RewardCampaignExtended> {
+        self.reward_campaigns
+            .iter()
+            .filter(|(_, campaign)| campaign.end_time > env::block_height())
+            .map(|(id, campaign)| RewardCampaignExtended {
+                campaign_id: id,
+                campaign: campaign.clone(),
+                market_total: self.get_market_total(campaign.clone()),
+                rewards_per_day: self.get_reward_tokens_per_day(campaign),
+            })
+            .collect::<Vec<RewardCampaignExtended>>()
+    }
+
+    pub fn get_rewards_per_time(&self, campaign: RewardCampaign, seconds: u128) -> WBalance {
+        let rewards_per_second: u128 = campaign.reward_amount.0 * seconds
+            / u128::from(campaign.end_time - campaign.start_time);
+        WBalance::from(rewards_per_second)
+    }
+
+    pub fn get_rewards_per_second(&self, campaign: RewardCampaign) -> WBalance {
+        self.get_rewards_per_time(campaign, 1)
+    }
+
+    pub fn get_reward_tokens_per_day(&self, campaign: RewardCampaign) -> WBalance {
+        self.get_rewards_per_time(campaign, 24 * 60 * 60)
+    }
+
+    pub fn get_market_total(&self, campaign: RewardCampaign) -> WBalance {
+        let total_amount = match campaign.campaign_type {
+            CampaignType::Supply => self.get_total_supplies(), // Tokens
+            CampaignType::Borrow => self.get_total_borrows(),  // Tokens
+        };
+        WBalance::from(total_amount)
+    }
+
+    pub fn get_account_total(&self, campaign: RewardCampaign, account_id: AccountId) -> WBalance {
+        let account_total = match campaign.campaign_type {
+            //TODO: uncomment get_account_supplies()
+            CampaignType::Supply => 0, // self.get_account_supplies(account_id), // Tokens
+            CampaignType::Borrow => self.get_account_borrows(account_id), // Tokens
+        };
+        WBalance::from(account_total)
+    }
+
+    fn insert_map_if_not_exists(&mut self, account_id: AccountId) {
+        let rewards_map: HashMap<String, Reward> = HashMap::new();
+        *self.rewards.entry(account_id).or_default() = rewards_map;
+    }
+
+    pub fn get_accrued_rewards_per_token(&self, campaign_id: String) -> WBalance {
+        if let Some(campaign) = self.get_reward_campaign_by_id(campaign_id.clone()) {
+            let reward_rate = self.get_rewards_per_second(campaign.clone());
+            let total = self.get_market_total(campaign.clone());
+            let current_time = min(env::block_timestamp(), campaign.end_time);
+            // println!(
+            //     "current_time {}, last_update_time {}, reward_rate {}, total {}",
+            //     current_time, campaign.last_update_time, reward_rate.0, total.0
+            // );
+            if total.0 == 0 {
+                return WBalance::from(0);
+            };
+            return WBalance::from(
```

Comment:
@pahomas we have accuracy loss here

current implementation unfolds into following expression
```sec_since_last_update * reward_rate / total``` ==>
```(current_time - campaign.last_update_time) * (campaign.reward_amount.0 / u128::from(campaign.end_time - campaign.start_time)) / total```

which has less accuracy than the following expression
```
(current_time - campaign.last_update_time) * campaign.reward_amount / (u128::from(campaign.end_time - campaign.start_time) / total
```

that happens because 1st case we do integer division and then multiplication and in 2nd case we first multiply and only then divide (take a look on how parenthesis are placed)